cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,270 @@
+      switch (__status)
+      {
+        case ::cudaSuccess:
+          break;
+        case ::cudaErrorInvalidValue:
+          return false;
+        default:
+          ::cudaGetLastError(); // Clear CUDA error state
+          ::cuda::__throw_cuda_error(__status, "Failed to call cudaDeviceGetAttribute");
+      }
+    }
+#    endif //_CCCL_CUDACC_BELOW_11_3
+    return (static_cast<int>(__handle_type) & __supported_handles) == static_cast<int>(__handle_type);
+  }
+
+  //! @brief  Creates the CUDA memory pool from the passed in arguments.
+  //! @throws cuda_error If the creation of the CUDA memory pool failed.
+  //! @returns The created CUDA memory pool.
+  _CCCL_NODISCARD static cudaMemPool_t
+  __create_cuda_mempool(const int __device_id, cuda_memory_pool_properties __properties) noexcept
+  {
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+    _LIBCUDACXX_ASSERT(__cuda_supports_export_handle_type(__device_id, __properties.allocation_handle_type),
+                       "Requested IPC memory handle type not supported");
+
+    ::cudaMemPoolProps __pool_properties{};
+    __pool_properties.allocType     = ::cudaMemAllocationTypePinned;
+    __pool_properties.handleTypes   = ::cudaMemAllocationHandleType(__properties.allocation_handle_type);
+    __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+    __pool_properties.location.id   = __device_id;
+    ::cudaMemPool_t __cuda_pool_handle{};
+    _CCCL_TRY_CUDA_API(::cudaMemPoolCreate, "Failed to call cudaMemPoolCreate", &__cuda_pool_handle, &__pool_properties);
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolSetAttribute with cudaMemPoolReuseAllowOpportunistic",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolSetAttribute with cudaMemPoolAttrReleaseThreshold",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__properties.release_threshold);
+
+    // allocate the requested initial size to prime the pool.
+    // We need to use a new stream so we do not wait on other work
+    if (__properties.initial_pool_size != 0)
+    {
+      ::cuda::experimental::stream __temp_stream{__device_id};
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
```

Comment:
looks like we're using `cudaMallocAsync` even if the driver version is less than `__min_async_version`.

Reply:
Above we have
```cpp
_LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
```
which should throw an exception if cudaMallocAsync is not supported.